cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java

```diff
@@ -0,0 +1,151 @@
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ProgramReference;
+import io.cdap.cdap.spi.events.EventReader;
+import io.cdap.cdap.spi.events.EventResult;
+import io.cdap.cdap.spi.events.StartProgramEvent;
+import io.cdap.cdap.spi.events.StartProgramEventDetails;
+import java.util.Collection;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.apache.twill.api.RunId;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Periodically checks event reader for StartProgramEvents. If any events available, publish them to
+ * TMS.
+ */
+public class StartProgramEventSubscriber extends EventSubscriber {
+
+  private static final Logger LOG = LoggerFactory.getLogger(StartProgramEventSubscriber.class);
+
+  private final CConfiguration cConf;
+  private final EventReaderProvider<StartProgramEvent> extensionProvider;
+  private final ProgramLifecycleService lifecycleService;
+  private ScheduledExecutorService executor;
+  private Collection<EventReader<StartProgramEvent>> readers;
+  private ExecutorService threadPoolExecutor;
+
+  /**
+   * Create instance that handles StartProgramEvents.
+   *
+   * @param cConf CDAP configuration
+   * @param extensionProvider eventReaderProvider for StartProgramEvent Readers
+   * @param lifecycleService to publish start programs to TMS
+   */
+  @Inject
+  StartProgramEventSubscriber(CConfiguration cConf, EventReaderProvider<StartProgramEvent> extensionProvider,
+                              ProgramLifecycleService lifecycleService) {
+    this.cConf = cConf;
+    this.extensionProvider = extensionProvider;
+    this.lifecycleService = lifecycleService;
+  }
+
+  @Override
+  public boolean initialize() {
+    readers = extensionProvider.loadEventReaders().values();
+    if (readers.isEmpty()) {
+      return false;
+    }
+    threadPoolExecutor = new ThreadPoolExecutor(readers.size(), readers.size(), 60,
+            TimeUnit.SECONDS, new LinkedBlockingQueue<>());
+    for (EventReader reader : readers) {
+      reader.initialize(
+              new DefaultEventReaderContext(Constants.Event.START_EVENT_PREFIX + "."
+                      + reader.getClass().getSimpleName(), cConf));
+    }
+    return true;
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedDelaySchedule(0,
+            cConf.getInt(Constants.Event.START_PROGRAM_EVENT_READER_POLL_DELAY), TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    if (threadPoolExecutor != null) {
+      for (EventReader<StartProgramEvent> reader : readers) {
+        threadPoolExecutor.execute(() -> {
+          EventResult<StartProgramEvent> result = reader.pull(1);
```

Comment:
why pull 1? It should be configurable with default value of 1.

Reply:
Should this be in CConf?